### PR TITLE
asttoc

### PR DESCRIPTION
**Pull request by luke-titley**
This is the working draft pull request for ast -> c binary.
Not much to see here right now, but progress will be tracked here.

So far @anderslanglands :
* I've copied the structs that define the structure of your ast into a header. cppmm_ast. I'm fine to keep those up to date myself with whatever you change in cppmm astdump branch. That gives me a stable structure that I can sync intermittently. Then eventually i'll just use whatever header you make.
* I've made a subdirectory asttoc. This is because I needed to override some of the settings in the CMakeLists.txt and I figured keeping the overrides scoped to a subdirectory was neater.
* I'm decoupling the serialization/deserialization from the ast nodes themselves, (bit like you had before), because I'm imagining transforming the ast directly (maybe adding a couple more structs for function body) then doing a c deserialisation. I think that's going to be the easiest way to write it so that it's maintainable. 

Also I messed around with cereal, but I'm going to move to nlohmann/json as well. I'm keen to remove cereal from the git history, so I'll be re-structuring the history to remove it before I ever un-draft this PR .

### CODE REVIEW

File: asttoc/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+cmake_minimum_required(VERSION 3.5)
+
+set(LLVM_ENABLE_RTTI ON)
```
Comment (anderslanglands) [inline asttoc/CMakeLists.txt:3]: How come you're enabling RTTI here? I'm building LLVM with RTTI off as that's what OSL uses.

Comment (luke-titley) [reply asttoc/CMakeLists.txt:3]: It's fine to remove this now. I originally planned on using cereal to read the json and cereal needed typeid, but I'm not using cereal anymore.

Comment (lgritz) [reply asttoc/CMakeLists.txt:3]: FYI, current OSL no longer disables RTTI across the board, that was causing problems with other dependencies that turned out to *require* RTTI. Now we only disable it [for the one file that directly calls LLVM API calls](https://github.com/AcademySoftwareFoundation/OpenShadingLanguage/blob/master/src/liboslexec/CMakeLists.txt#L206). I isolate the LLVM calls in one wrapper class module, which doesn't need to call into any other dependencies that might use RTTI. You can make the RTTI on/off choice module-by-module, as it turns out.

Comment (lgritz) [reply asttoc/CMakeLists.txt:3]: My memory might be hazy. Now that I think about it, I believe it was not OSL's dependencies that were the problems, but downstream apps that needed to link against both OSL (which was building RTTI off across the board) and something else that required RTTI, and was having trouble with that mix. So I had to stop "exporting" the no-RTTI requirement from OSL, and instead restrict it to just the one module that directly called LLVM which is usually built without RTTI.

Comment (anderslanglands) [reply asttoc/CMakeLists.txt:3]: Ah that's good to know, thanks Larry.

**Pull request merged**